examples/sig-example.cpp

```diff
@@ -0,0 +1,198 @@
+{
+    std::shared_ptr<MessageVerificationCtx> ctx(nullptr);
+    try {
+        ctx = std::make_shared<ECDSASignaturePublicKeyCtx>(*pubKey, digestType, sigFormat);
+    }  catch (MoCOCrWException &e) {
+        std::cerr << "Please check your ECC key. Failure creating context." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+    return ctx;
+}
+
+std::shared_ptr<MessageSignatureCtx>getEdDsaSignCtx(std::shared_ptr<AsymmetricPrivateKey> privKey)
+{
+    std::shared_ptr<MessageSignatureCtx> ctx(nullptr);
+    try {
+        ctx = std::make_shared<EdDSASignaturePrivateKeyCtx>(*privKey);
+    }  catch (MoCOCrWException &e) {
+        std::cerr << "Please check your ECC-Ed key. Failure creating context." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+    return ctx;
+}
+
+std::shared_ptr<MessageVerificationCtx> getEdDsaVerifyCtx(std::shared_ptr<AsymmetricPublicKey> pubKey)
+{
+    std::shared_ptr<MessageVerificationCtx> ctx(nullptr);
+    try {
+        ctx = std::make_shared<EdDSASignaturePublicKeyCtx>(*pubKey);
+    }  catch (MoCOCrWException &e) {
+        std::cerr << "Please check your ECC-Ed key. Failure creating context." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+    return ctx;
+}
+
+std::vector<uint8_t> sign(std::shared_ptr<MessageSignatureCtx> signCtx,
+                          const std::vector<uint8_t> &message)
+{
+    std::vector<uint8_t> signature;
+    try {
+        signature = signCtx->signMessage(message);
+    }  catch (MoCOCrWException &e) {
+        std::cerr << "Failure occurred during signing." << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+    return signature;
+}
+
+void verify(std::shared_ptr<MessageVerificationCtx> verifyCtx,
+            const std::vector<uint8_t> &message,
+            const std::vector<uint8_t> &signature)
+{
+    try {
+        verifyCtx->verifyMessage(signature, message);
+    } catch (MoCOCrWException &e) {
+        std::cerr << "Verification failed!" << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
+
+
+int main(void) {
+    DigestTypes digestType = DigestTypes::SHA512;
+    std::vector<uint8_t> message = utility::fromHex("deadbeef");
+
+    /************** RSA signature **************/
+    auto rsaPrivKey = std::make_shared<AsymmetricPrivateKey>(AsymmetricPrivateKey::generateRSA());
+    auto mgf1 = std::make_shared<MGF1>(MGF1(DigestTypes::SHA256));
+    auto padding = std::make_shared<PSSPadding>(PSSPadding(mgf1, mococrw::Hash::getDigestSize(digestType)));
+
+    auto rsaSignCtx = getRsaSignCtx(rsaPrivKey, digestType, padding);
+    auto signature = sign(rsaSignCtx, message);
+
+    /* we can use here the private key, as it also contains the public key.
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
+     * we do an implicit downcast here.
```

Comment:
Isn't this an upcast, since we're casting to the superclass (i.e. the one above in the type hierarchy)?